File: entity_extractor.py
"""
Entity Extraction Module
Named Entity Recognition using spaCy and OpenAI
"""

import json
from typing import List, Dict, Optional
import spacy
from config import SPACY_MODEL
from openai_helper import OpenAIHelper


class EntityExtractor:
    """Extract named entities using spaCy and OpenAI"""
    
    def __init__(self, openai_api_key: Optional[str] = None):
        self.nlp = spacy.load(SPACY_MODEL)
        self.openai = OpenAIHelper(openai_api_key)
        self.use_llm = bool(openai_api_key)
    
    def extract_entities_spacy(self, text: str) -> List[Dict]:
        """Extract entities using spaCy"""
        doc = self.nlp(text)
        entities = []
        
        for ent in doc.ents:
            entities.append({
                'text': ent.text,
                'label': ent.label_,
                'start': ent.start_char,
                'end': ent.end_char
            })
        
        print(f"✓ Extracted {len(entities)} entities using spaCy")
        return entities
    
    def extract_entities_llm(self, text: str) -> List[Dict]:
        """Extract entities using OpenAI LLM for better accuracy"""
        if not self.use_llm:
            return []
        
        prompt = f"""Extract all named entities from the following text. 
        Return them as a JSON list with format: [{{"text": "entity", "type": "PERSON/ORG/GPE/DATE/WORK_OF_ART/etc"}}]
        
        Text: {text}
        
        Entities:"""
        
        messages = [
            {"role": "system", "content": "You are an expert at named entity recognition. Return only valid JSON."},
            {"role": "user", "content": prompt}
        ]
        
        response = self.openai.chat_completion(messages)
        
        if not response:
            return []
        
        try:
            content = response.strip()
            if "```" in content:
                content = content.split("```")[1]
                if content.startswith("json"):
                    content = content[4:]
            
            entities = json.loads(content)
            print(f"✓ Extracted {len(entities)} entities using LLM")
            return entities
        except Exception as e:
            print(f"✗ LLM entity extraction failed: {e}")
            return []
    
    def merge_entities(self, spacy_entities: List[Dict], llm_entities: List[Dict]) -> List[Dict]:
        """Merge and deduplicate entities from different sources"""
        entity_map = {}
        
        for ent in spacy_entities:
            key = ent['text'].lower()
            entity_map[key] = {
                'text': ent['text'],
                'type': ent['label']
            }
        
        for ent in llm_entities:
            key = ent['text'].lower()
            entity_map[key] = {
                'text': ent['text'],
                'type': ent.get('type', 'UNKNOWN')
            }
        
        entities = list(entity_map.values())
        print(f"✓ Merged into {len(entities)} unique entities")
        return entities

File: backend/src/main/kotlin/io/github/juantoanxoup/kg/EntityExtractor.kt
package io.github.juantoanxoup.kg

import ai.koog.prompt.dsl.prompt
import ai.koog.prompt.executor.model.PromptExecutor
import ai.koog.prompt.executor.model.StructureFixingParser
import ai.koog.prompt.executor.model.executeStructured
import ai.koog.prompt.params.LLMParams
import edu.stanford.nlp.pipeline.CoreDocument
import kotlinx.serialization.Serializable
import org.slf4j.LoggerFactory

/** Entity after merge: surface text and coarse type. */
@Serializable
data class Entity(
    val text: String,
    val type: String,
)

/** Entity from the NLP pass with character offsets. */
@Serializable
data class NlpEntity(
    val text: String,
    val label: String,
    val start: Int,
    val end: Int,
)

/** Structured output requested from the LLM. */
@Serializable
data class ExtractedEntities(
    val entities: List<ExtractedEntity>,
)

@Serializable
data class ExtractedEntity(
    val text: String,
    val type: String? = null,
)

/** Named entity recognition with CoreNLP and an LLM (entity_extractor.py). */
class EntityExtractor(
    private val executor: PromptExecutor? = OpenAiHelper.promptExecutor(),
) {
    private val log = LoggerFactory.getLogger(EntityExtractor::class.java)

    /** NLP pass. CoreNLP labels are mapped to the label set the graph builder understands. */
    fun extractEntitiesNlp(text: String): List<NlpEntity> {
        val doc = CoreDocument(text)
        Nlp.pipeline.annotate(doc)
        val entities =
            doc.entityMentions().map { mention ->
                val offsets = mention.charOffsets()
                NlpEntity(
                    text = mention.text(),
                    label = mapLabel(mention.entityType()),
                    start = offsets.first(),
                    end = offsets.second(),
                )
            }
        log.info("Extracted {} entities using CoreNLP", entities.size)
        return entities
    }

    /**
     * LLM pass over the whole text as a structured-output request. Returns an empty list when no
     * executor is configured or the response cannot be parsed even after Koog's fixing pass.
     */
    suspend fun extractEntitiesLlm(text: String): List<Entity> {
        val llm = executor ?: return emptyList()
        val request =
            prompt("entity-extraction", LLMParams(temperature = 0.0)) {
                system("You are an expert at named entity recognition. Return only valid JSON.")
                user(
                    """
                    Extract all named entities from the following text.
                    Give each entity a type such as PERSON, ORG, GPE, DATE, WORK_OF_ART, EVENT.

                    Text: $text
                    """.trimIndent(),
                )
            }
        val result =
            llm.executeStructured<ExtractedEntities>(
                prompt = request,
                model = OpenAiHelper.defaultModel,
                fixingParser = StructureFixingParser(OpenAiHelper.defaultModel, retries = 1),
            )
        return result
            .map { response -> response.data.entities.map { Entity(it.text, it.type ?: "UNKNOWN") } }
            .onSuccess { log.info("Extracted {} entities using LLM", it.size) }
            .onFailure { log.warn("LLM entity extraction failed: {}", it.message) }
            .getOrDefault(emptyList())
    }

    /** Keyed by lower-cased text; LLM entities overwrite NLP entities on collision. */
    fun mergeEntities(
        nlpEntities: List<NlpEntity>,
        llmEntities: List<Entity>,
    ): List<Entity> {
        val entityMap = LinkedHashMap<String, Entity>()
        for (e in nlpEntities) entityMap[e.text.lowercase()] = Entity(e.text, e.label)
        for (e in llmEntities) entityMap[e.text.lowercase()] = Entity(e.text, e.type)
        val entities = entityMap.values.toList()
        log.info("Merged into {} unique entities", entities.size)
        return entities
    }

    private companion object {
        /** CoreNLP coarse labels to the spaCy-style labels used by the graph builder. */
        fun mapLabel(label: String): String =
            when (label) {
                "ORGANIZATION" -> "ORG"
                "LOCATION", "CITY", "COUNTRY", "STATE_OR_PROVINCE" -> "GPE"
                else -> label
            }
    }
}
